lib/virt.py

```diff
@@ -596,21 +603,33 @@ def is_installed(self):
         tag = self.install_ready_path.read_text()
         return tag == self.tag
 
-    def can_be_snapshotted(self):
-        if not self.snapshot_ready_path.exists():
-            return False
-        tag = self.snapshot_ready_path.read_text()
-        return tag == self.tag
-
     def prepare_for_snapshot(self):
-        # do guest first boot, let it settle and finish firstboot tasks,
-        # then shut it down + start again, to get the lowest possible page cache
-        # (resulting in smallest possible RAM image)
+        if not self.is_installed():
+            raise RuntimeError(f"guest {self.name} not installed or installed with different tag")
+
+        # if an external domain is used (not one we installed) and it is in inconsistent state
+        # (using snapshot-style disk), shut it down and restore its original RAM and disk image
+        if not self.disk_path:
+            current_disk_path, _ = get_state_image_disk(self.state_file_path)
+            self.disk_path, self.disk_format = get_state_base_image_disk(self.name)
+            if current_disk_path != self.disk_path:
+                util.log(f"found {self.name} without reverted snapshot")
+                util.log(
+                    f"restoring {self.name} from {self.state_file_path}, "
+                    f"using original base image disk: {self.disk_path}",
+                )
+                self.destroy()
+                set_state_image_disk(self.state_file_path, self.disk_path, self.disk_format)
+                virsh('restore', self.state_file_path, check=True)
+
+        # do guest first boot, let it settle and finish firstboot tasks
         self.start()
-        ip = wait_for_ifaddr(self.name)
-        wait_for_ssh(ip)
+        self.ipaddr = wait_for_ifaddr(self.name)
+        wait_for_ssh(self.ipaddr)
```

Comment:
I'm not sure this is a good idea, we already do this in both `booted()` and `snapshotted()` later, and duplicating it here is IMHO somewhat confusing.

Maybe we could do it only here and not in `snapshotted()`, since we now ALWAYS need to `prepare_for_snapshot()` before using `snapshotted()`?